crates/data_components/src/databricks/sql_warehouse.rs

```diff
@@ -0,0 +1,634 @@
+    > {
+        Ok(Box::new(SqlWarehouseConnection {
+            api: Arc::clone(&self.api),
+        }))
+    }
+
+    fn join_push_down(&self) -> JoinPushDown {
+        JoinPushDown::Disallow
+    }
+}
+
+struct SqlWarehouseApi {
+    client: Client,
+    host: String,
+    sql_warehouse_id: String,
+    token_provider: Arc<dyn TokenProvider>,
+}
+
+impl SqlWarehouseApi {
+    fn new(
+        host: &str,
+        sql_warehouse_id: &str,
+        token_provider: Arc<dyn TokenProvider>,
+    ) -> Result<Self, Error> {
+        let client = ClientBuilder::new()
+            .user_agent(super::user_agent())
+            .build()
+            .context(ClientBuildFailedSnafu)?;
+
+        Ok(Self {
+            client,
+            host: host.to_string(),
+            sql_warehouse_id: sql_warehouse_id.to_string(),
+            token_provider,
+        })
+    }
+
+    async fn get_schema(&self, table: &TableReference) -> Result<SchemaRef, Error> {
+        let token = self.token_provider.get_token();
+        let sql = format!("DESCRIBE TABLE {table}");
+        let payload = self.create_schema_payload(table, &sql)?;
+
+        let response = self.execute_request(&token, &payload).await?;
+        schema_from_json(&response)
+    }
+
+    fn create_schema_payload(&self, table: &TableReference, sql: &str) -> Result<Value, Error> {
+        Ok(json!({
+            "warehouse_id": self.sql_warehouse_id,
+            "catalog": table.catalog().ok_or_else(|| Error::FullyQualifiedPath{ reason: "missing catalog".into() })?,
+            "schema": table.schema().ok_or_else(|| Error::FullyQualifiedPath{ reason: "missing schema".into() })?,
+            "statement": sql,
+        }))
+    }
+
+    async fn execute_request(&self, token: &str, payload: &Value) -> Result<Value, Error> {
+        let url = format!("https://{}/api/2.0/sql/statements/", self.host);
+        self.client
+            .post(&url)
+            .bearer_auth(token)
+            .json(payload)
+            .send()
+            .await
+            .context(HttpRequestFailedSnafu)?
+            .json()
+            .await
+            .context(JsonParsingFailedSnafu)
+    }
+
+    // Fetch the arrow data at the external links, repeating for each chunk
+    async fn fetch_external_links(
+        &self,
+        result_object: Value,
+    ) -> Result<SendableRecordBatchStream, Error> {
+        let mut streams = Vec::new();
+        let token = self.token_provider.get_token();
+        let mut external_link = Self::extract_external_links(result_object)?;
+        loop {
+            tracing::trace!(
+                "Fetching chunk {} from {}",
+                external_link.chunk_index,
+                external_link.external_link
+            );
+            let bytes = self.fetch_chunk_data(&external_link.external_link).await?;
+            let batches = Self::read_arrow_batches(bytes)?;
+            streams.push(futures::stream::iter(batches.into_iter().map(Ok)));
```

Comment:
This is buffering all of the batches into memory before returning the stream. It should return the batches into the stream as it is processing. So this logic to fetch the data should be in the stream itself.